src/Features/CSharp/Portable/Completion/CompletionProviders/PropertySubPatternCompletionProvider.cs

```diff
@@ -0,0 +1,199 @@
+                return default;
+            }
+
+            return token;
+        }
+
+        private static ITypeSymbol TryGetPatternType(
+            SyntaxToken token, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            // is Goo { $$
+            // is Goo { P1: 0, $$
+            var propertyPattern = (PropertyPatternSyntax)token.Parent.Parent;
+            var patternType = propertyPattern.Type;
+            if (patternType != null)
+            {
+                // TODO test alias
+                var typeSymbol = (ITypeSymbol)semanticModel.GetSymbolInfo(patternType, cancellationToken).Symbol;
+                return typeSymbol;
+            }
+
+            // e is { $$
+            // e is { P1: 0, $$
+            if (propertyPattern.IsParentKind(SyntaxKind.IsPatternExpression))
+            {
+                var isPattern = (IsPatternExpressionSyntax)propertyPattern.Parent;
+                var typeSymbol = semanticModel.GetTypeInfo(isPattern.Expression, cancellationToken).Type;
+                if (typeSymbol != null)
+                {
+                    return typeSymbol;
+                }
+            }
+
+            // switch (e) { case { $$
+            // switch (e) { case { P1: 0, $$
+            if (propertyPattern.IsParentKind(SyntaxKind.CasePatternSwitchLabel))
+            {
+                var casePattern = (CasePatternSwitchLabelSyntax)propertyPattern.Parent;
+                var switchStatment = casePattern.Parent?.Parent as SwitchStatementSyntax;
+                if (switchStatment != null && switchStatment.Expression != null)
+                {
+                    var typeSymbol = semanticModel.GetTypeInfo(switchStatment.Expression).Type;
+                    if (typeSymbol != null)
+                    {
+                        return typeSymbol;
+                    }
+                }
+            }
+
+            // e is { P1: { $$
+            // e is { P1: { P2: 2, $$
+            if (propertyPattern.IsParentKind(SyntaxKind.SubpatternElement))
+            {
+                var containingProperty = ((SubpatternElementSyntax)propertyPattern.Parent).NameColon?.Name;
+                if (containingProperty != null)
+                {
+                    var symbol = semanticModel.GetSymbolInfo(containingProperty).Symbol;
+
+                    switch (symbol)
+                    {
+                        case null:
+                            break;
+                        case IPropertySymbol propertySymbol:
+                            return propertySymbol.Type;
+                        case IFieldSymbol fieldSymbol:
+                            return fieldSymbol.Type;
+                    }
+                }
+            }
+
+            // PROTOTYPE(NullableReferenceTypes): 
+            // All the cases above should be replaced by using the semantic model's GetTypeInfo (ConvertedType) once that is integrated
+
+            return default;
```

Comment:
i feel like there's a bunch of a logic shared here and between the work i'm doing in my PR where i update the TypeInferenceService...  i don't have a great suggestion on how to unify things though :-/

Reply:
The logic for ` TryGetPatternType` can be removed once Neal's [PR for GetTypeInfo](https://github.com/dotnet/roslyn/pull/26521) gets merged. 

There will still be some holes. For instance, when typing a property pattern inside a positional pattern and the positional pattern isn't completely typed yet, so the `Deconstruct` method may not be fully resolved yet. But I think we'll fix that from compiler side (best overload candidate should be returned in that case, for IDE to use (just as in SignatureHelp).

Beyond that, I don't expect much need for inference from IDE side in patterns.